designs/2023-only-run-reporting-rules/README.md

```diff
@@ -0,0 +1,129 @@
+
+## Motivation
+
+A common use-case of the `warn` level is to report errors to the developer in their IDE,
+but to completely ignore it when running ESLint on the command line via the `--quiet` flag.
+When running with many rules set to the `warn` level, this can cause a significant amount
+of overhead on ESLint runtime. On an example large codebase, I found that by disabling all
+rules set to warn, I was able to reduce ESLint runtime from 4 minutes and 39 seconds to just
+22 seconds. As the `warn` rules are entirely ignored in this situation, this is a substantial
+amount of wasted time and resources that could be avoided.
+
+## Detailed Design
+
+The proposed implementation is to add a `--skip-warnings` flag to ESLint that will skip
+running any rules that are set to the `warn` level. This will be a CLI flag only, as it
+doesn't make sense to have this flag in a configuration file. This flag would implicitly
+imply the `--quiet` flag, given if the `warn` rules are not run, they will not be outputted
+to reporters.
+
+From an API perspective, this would be implemented by a predicate function that acts as
+a filter on which rules should be run. The predicate would take the rule configuration,
+and return whether the rule should be run. For simplicity, this function should not be
+given rules marked as `off`, as if this function handled existing behavior then users of the
+API would have to mimic that when attempting to extend it.
+
+The flag would be implemented by passing a predicate function that only returns true for
+rules that have been set to `error`, thus filtering out any that are marked as `warn`.
+
+The check for unused `eslint-disable` directives (`--report-unused-disable-directives`)
+should continue to mark `warn` rules as used, even when running with `--skip-warnings`.
+
+The `--max-warnings` flag should disable the `--skip-warnings` flag, as it requires that
+`warn` rules are run to know the current number of reported warnings.
+
+## Documentation
+
+Both the flag and API would make sense to document. As they are relatively simple additions,
+they would be documented inline with the rest of the CLI and API documentation. Due to the
+potentially major performance improvements for many setups, I feel mentioning it in the
+release blog post, alongside the potential benefits, would be a good idea to spread awareness.
+
+## Drawbacks
+
+This change adds a new command line flag, as well as a new API method. This adds further
+maintenance burden. However, the implementation is relatively simple, and the benefits
+are significant.
+
+Another potential drawback around this specific implementation is that the existence of
+both `--quiet` and `--skip-warnings` might be confusing to some users, however this can be
+partially alleviated by updating the `--quiet` docs to mention that warnings will still be
+run, and to use `--skip-warnings` instead to not run them.
```

Comment:
It's quite hard for me to guess the difference between `--quiet` and a hypothetical `--skip-warnings` option just from their names. While the docs could explain this, I think it's a sign that the design and naming may not be intuitive to users.